service/resource/endpoint/create.go

```diff
@@ -1,11 +1,133 @@
 package endpoint
 
-import "context"
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiv1 "k8s.io/client-go/pkg/api/v1"
+)
 
 func (r *Resource) GetCreateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error) {
-	return nil, nil
+	currentEndpoints, err := toEndpointSlice(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredEndpoints, err := toEndpointSlice(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var createState = []Endpoint{}
+	for _, desiredEndpoint := range desiredEndpoints {
+		if !containsEndpoint(currentEndpoints, desiredEndpoint) {
+			createState = append(createState, desiredEndpoint)
+		}
+	}
+
+	if len(createState) == 0 {
+		return nil, nil
+	}
+
+	return createState, nil
 }
 
 func (r *Resource) ProcessCreateState(ctx context.Context, obj, createState interface{}) error {
+	endpointsToCreate, err := toEndpointSlice(createState)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	for _, endpoint := range endpointsToCreate {
+		k8sService, err := r.k8sClient.CoreV1().Services(endpoint.ServiceNamespace).Get(endpoint.ServiceName, metav1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		k8sEndpoint := &apiv1.Endpoints{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: endpoint.ServiceName,
+			},
+			Subsets: []apiv1.EndpointSubset{
+				{
+					Addresses: []apiv1.EndpointAddress{
+						{
+							IP: endpoint.IP,
+						},
+					},
+					Ports: serviceToPorts(k8sService),
+				},
+			},
+		}
+		_, err = r.k8sClient.CoreV1().Endpoints(endpoint.ServiceNamespace).Create(k8sEndpoint)
+		if errors.IsAlreadyExists(err) {
+			// In case the endpoint we tried to create does already exist, we only
+			// need to append the IPs we got in podInfos.
+			err = appendIPs(r, endpoint.ServiceNamespace, endpoint.ServiceName, endpoint.IP)
```

Comment:
AFAIK the endpoint will always exist. We fetched the current state from the endpoint. In case we cannot find it we cannot proceed with the reconciliation anyway.

Reply:
Yes I removed this.